### PR TITLE
Combining all the evaluation code

### PR DESCRIPTION
**Pull request by lizgzil**
# Description
Now that all the PRs for each part of the evaluation have been merged, there was an error with a variable name (due to me not doing a merge conflict properly) and there were some general things to neaten up.

In this PR I:
- Neaten up code
- Correct the variable name in the parse test
- Add to document describing the evaluation data and the metrics returned (now renamed).
- Rename all mentions of 'test' to 'evaluation' or 'metric'

# Results
Note:
- evaluation 1 and 2 still don't have the actual evaluation data.
- you may wonder why the evaluation 3 results are different from the recent work we did on the references splitting since this is the same evaluation - it's because that was with the IOBE model, whereas the current function for `split_section` in policytool still uses the regex. We should improve evaluation 3 to 77.4 when we include the IOBE line model scraper (https://github.com/wellcometrust/datalabs/tree/master/modelling/policy/reference-splitter).
- the levenshtein thresholds of 0.3 are quite arbitrary.

verbose = False
```
Date of evaluation = 2019-05-24-1546

Score for evaluation 1:
0.7

Score for evaluation 2:
0.0

Score for evaluation 3:
52.8

Score for evaluation 4:
0.4

Score for evaluation 5:
1.0

```

verbose = True
```
Date of evaluation = 2019-05-24-1545

-----Information about evaluation 1:-----

Score
0.7

F1-score
0.7

Classification report
              precision    recall  f1-score   support

       False       0.67      1.00      0.80         6
        True       1.00      0.25      0.40         4

   micro avg       0.70      0.70      0.70        10
   macro avg       0.83      0.62      0.60        10
weighted avg       0.80      0.70      0.64        10


Confusion matrix
                Predicted True  Predicted False
Actually True                1                3
Actually False               0                6

Classification report for the reference section
              precision    recall  f1-score   support

       False       0.50      1.00      0.67         2
        True       1.00      0.33      0.50         3

   micro avg       0.60      0.60      0.60         5
   macro avg       0.75      0.67      0.58         5
weighted avg       0.80      0.60      0.57         5


Confusion matrix for the reference section
                Predicted True  Predicted False
Actually True                1                2
Actually False               0                2

Classification report for the bibliograph section
              precision    recall  f1-score   support

       False       0.80      1.00      0.89         4
        True       0.00      0.00      0.00         1

   micro avg       0.80      0.80      0.80         5
   macro avg       0.40      0.50      0.44         5
weighted avg       0.64      0.80      0.71         5


Confusion matrix for the bibliograph section
                Predicted True  Predicted False
Actually True                0                1
Actually False               0                4

-----Information about evaluation 2:-----

Score
0.0

Mean normalised Levenshtein distance
0.8981481481481481

Strict accuracy (micro)
0.0

Lenient accuracy (micro)
0.0

Mean normalised Levenshtein distance for the reference section
0.8641975308641975

Strict accuracy for the reference section
0.0

Lenient accuracy for the reference section
0.0

Mean normalised Levenshtein distance for the bibliograph section
1.0

Strict accuracy for the bibliograph section
0.0

Lenient accuracy for the bibliograph section
0.0

-----Information about evaluation 3:-----

Score
52.8

Percentage below threshold of 40
52.8

Median difference
29.6

Metrics grouped by source
          Median difference  Percentage below threshold of 40
Source                                                       
msf                     7.3                              83.3
nice                   95.8                               0.0
unicef                 27.4                              90.0
who_iris               10.0                              82.4

-----Information about evaluation 4:-----

Score
0.402787456445993

Strict accuracy (micro)
0.402787456445993

Strict accuracy (per category)
Authors       0.341463
Issue         0.639024
Journal       0.336585
Pagination    0.546341
PubYear       0.292683
Title         0.243902
Volume        0.419512
dtype: float64

Lenient accuracy (micro)        (normalised Levenshtein < 0.3)
0.8167247386759582

Lenient accuracy (per category)        (normalised Levenshtein < 0.3) ()
Authors       0.614634
Issue         0.917073
Journal       0.839024
Pagination    0.912195
PubYear       0.956098
Title         0.609756
Volume        0.868293
dtype: float64

Mean normalised Levenshtein distance (all categories)
0.13793400430184963

Mean normalised Levenshtein distance (per category)
Authors       0.299441
Issue         0.074099
Journal       0.136707
Pagination    0.055345
PubYear       0.036277
Title         0.256026
Volume        0.107641
dtype: float64

-----Information about evaluation 5:-----

Score
0.998

Micro average F1-score
0.998

Classification report
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       202
           1       1.00      0.99      1.00       199

   micro avg       1.00      1.00      1.00       401
   macro avg       1.00      1.00      1.00       401
weighted avg       1.00      1.00      1.00       401


Frequency table of match types
Matched correctly?     False  True  No match found
Do we expect a match?                             
0                         15     0             187
1                          1   198               0

```

### CODE REVIEW

File: policytool/refparse/algo_evaluation/evaluate_split_section.py [hunk 1]
```diff
@@ -60,20 +60,20 @@ def evaluate_metrics(actual_pred_num_references, threshold):
...
     actual_pred_num_references = []
-    for references_section in split_section_test_data['Reference section']:
+    for references_section in evaluate_split_section_data['Reference section']:
         references = split_section(references_section, regex = regex)
         actual_pred_num_references.append({
             "Predicted references" : references,
             "Predicted number of references" : len(references)
         })
-    actual_pred_num_references = pd.concat([split_section_test_data, pd.DataFrame(actual_pred_num_references)], axis = 1)
+    actual_pred_num_references = pd.concat([evaluate_split_section_data, pd.DataFrame(actual_pred_num_references)], axis = 1)
```
Comment (nsorros) [inline policytool/refparse/algo_evaluation/evaluate_split_section.py]: I am generally against variable names that start with a verb, i reserve those for functions.

File: policytool/refparse/evaluate_algo.py [hunk 2]
```diff
@@ -40,17 +40,17 @@ def yield_section_data(scrape_pdf_location, sections_location):
...
 
 def create_argparser():
     parser = ArgumentParser()
     parser.add_argument(
         '--verbose',
-        help='Whether you want to print detailed test \
+        help='Whether you want to print detailed evaluation \
             information ("True") or not ("False")',
-        default = True
+        default = 'True'
```
Comment (nsorros) [inline policytool/refparse/evaluate_algo.py]: why did you change that to be a string?

File: policytool/refparse/evaluate_algo.py [hunk 3]
```diff
@@ -151,46 +151,58 @@ def create_argparser():
...
+        eval1_scores,
+        eval2_scores,
+        eval3_scores,
+        eval4_scores,
+        eval5_scores
+        ]
 
-    if args.verbose:
-        for i, tests in enumerate(test_scores_list):
+    if eval(args.verbose):
```
Comment (nsorros) [inline policytool/refparse/evaluate_algo.py]: If you do not use a string for representing args.verbose then you do not need to do eval

**Comment by nsorros** [comment]
When I run `python evaluate_algo.py` I get this error `FileNotFoundError: [Errno 2] No such file or directory: './algo_evaluation/data_evaluate/pdf_sections'`

File: docs/evaluation.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
+In this document we describe how we got each of the evaluation datasets uses in `evaluate_algo.py`. This data can be downloaded [here](https://s3-eu-west-1.amazonaws.com/datalabs-data/policy_tool_tests) and are stored in the repository [here](policytool/refparse/algo_evaluation/data_evaluate). We also describe the output when running `evaluate_algo`.
+
+
+## Find Sections Evaluation
+### Data
+- We generated a random list of policy documents we currently scrape from various policy organisations.
+- We went through this list until we found X which had a section that contain references and X that didn't.
+- The pdfs for these were saved in "./algo_evaluation/data_evaluate/pdfs", named with a unique identifier for this pdf.
+- We looked for the text from sections Reach currently looks for in the pdfs ('reference' and 'bibliograph'), stored [here](policytool/resources/section_keywords.txt).
+- The text was copied and pasted from the pdf into a markdown file and saved as a .txt file in the relevant sections folder (e.g. "./algo_evaluation/data_evaluate/pdf_sections/reference") and saved with the same unique name as the pdf was.
```
Comment (nsorros) [inline docs/evaluation.md]: It is not a markdown file

File: docs/evaluation.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+
+### Evaluation Scores 1 and 2
+- We use the functions `parse_pdf_document` and `grab_section` from `policytool/pdf_parser/pdf_parse.py` to predict the references sections from each of the pdfs in our evaluation.
```
Comment (nsorros) [inline docs/evaluation.md]: grab section is the star of this show. we should probable rename it as well.

File: docs/evaluation.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+2. In the pdfs which do have a references section, we find the normalised Levenshtein distances for each of the actual and predicted references section texts. The proportion of pdfs with exactly the same actual and predicted texts is returned - the 'Strict accuracy', and we also return a more lenient score of the proportion where texts are closely similar (this is defined by a threshold parameter - `LEVENSHTEIN_DIST_SCRAPER_THRESHOLD`) - the 'Lenient accuracy'.
+
+## Split Section Evaluation
+### Data
+- Randomly sample policy documents with scraped references sections.
+- Made a decision about the quality of the scrape - if it was bad then we wouldn't include, but if the references text scraped included the whole references text we took it forward. We have 6 documents from MSF documents, 20 from NICE, 10 from UNICEF, and 17 from WHO documents.
```
Comment (nsorros) [inline docs/evaluation.md]: Another way we could have isolated only good scrapes would be to clean the ones that were badly scraped not just skip them.

File: docs/evaluation.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+- We also return the proportion which are quite similarly predicted (using the parameter `LEVENSHTEIN_DIST_PARSE_THRESHOLD`) - the 'Lenient accuracy', and break down the results by category.
+
+
+## Match Evaluation
+
+### Data
+- We randomly selected 200 references from a list of Wellcome Trust publications from Uber Dimensions ("uber_api_publications.csv").
+- With this list (the 'positive' list) we used the fuzzy match algorithm to find a list (the 'negative' list) of the second most similar references to them (i.e. the best match would be the reference in question, and the second would be quite similar).
+- The positive list was reduced to 199 since one reference didn't have a second strong match, and the negative list gave 202 references since for some references there were multiple references with the same cosine similarity competing for second place.
+- We take the titles and publication IDs from these and combine them in "positive_and_negative_match_test_data.csv" where we record a 1 or 0 for the positive and negative lists respectively in 'Do we expect a match?'.
```
Comment (nsorros) [inline docs/evaluation.md]: where do we record 1 or 0, you do not mention which columns 1 and 0 go and what do they mean.

File: policytool/refparse/evaluate_algo.py (hunk 2 above)
Comment (lizgzil) [reply policytool/refparse/evaluate_algo.py]: It doesn't seem to deal well with boolean arguments.

e.g.
```
   parser.add_argument(
        '--verbose',
        help='Whether you want to print detailed evaluation \
            information ("True") or not ("False")',
        default = True,
        type=bool
    )
```
then
```
python -i evaluate_algo.py --verbose False
```
gives
```
>>> args.verbose
True
```

However, when I dont include the type=bool, i.e. 
```
    parser.add_argument(
        '--verbose',
        help='Whether you want to print detailed evaluation \
            information ("True") or not ("False")',
        default = 'True'
    )
```
then
```
python -i evaluate_algo.py --verbose False
```
gives
```
>>> args.verbose
'False'
```
So I just thought I'd deal with them as strings, seems easier to read than the solutions listed [here](https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse)
🤷‍♀

File: docs/evaluation.md (hunk 5 above)
Comment (lizgzil) [reply docs/evaluation.md]: To me `parse_pdf_document` isn't a basic function and might involve assumptions we should be evaluating (@SamDepardieu can help tell us about any assumptions he made). So shouldn't the evaluation be `parse_pdf_document` plus `grab_section`?

File: docs/evaluation.md (hunk 6 above)
Comment (lizgzil) [reply docs/evaluation.md]: true

File: docs/evaluation.md (hunk 7 above)
Comment (lizgzil) [reply docs/evaluation.md]: I've reworded this to 
- We take the titles and publication IDs from these and combine them in "positive_and_negative_match_test_data.csv", and in this create a 'Do we expect a match?' column where we record a '1' if this reference was from the positive list and a '0' if it was from the negative list.

does that make it clearer?

File: policytool/refparse/algo_evaluation/evaluate_split_section.py (hunk 1 above)
Comment (lizgzil) [reply policytool/refparse/algo_evaluation/evaluate_split_section.py]: You probably noticed, but all the data names are in the form _x_`_data`, where x is the name of the function where this data will be evaluated. 
I could do it the other way round? like `data_evaluate_split_section`?

File: README.md [hunk 8]
```diff
@@ -123,11 +123,17 @@ We have devised some evaluation data in order to evaluate 5 steps of the model.
 ```
 brew install poppler
 ```
-and then downloading the evaluation data from [here](https://s3-eu-west-1.amazonaws.com/datalabs-data/policy_tool_tests), and then running
+and then downloading the evaluation data from [here](https://s3-eu-west-1.amazonaws.com/datalabs-data/policy_tool_tests) and storing it in `algo_evaluation/data_evaluate/`, which can be done in the command line by running
+```
+aws s3 cp --recursive s3://datalabs-data/policy_tool_tests/pdf_sections algo_evaluation/data_evaluate/
```
Comment (nsorros) [inline README.md]: nice 👍

File: docs/evaluation.md (hunk 5 above)
Comment (nsorros) [reply docs/evaluation.md]: I do not think we will ever write our own parser but I guess you are referring to some parameters we might tweak. Fair enough, let's leave it for now. I was just voicing what I had in my head we were doing.

File: docs/evaluation.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+## Match Evaluation
+
+### Data
+- We randomly selected 200 references from a list of Wellcome Trust publications from Uber Dimensions ("uber_api_publications.csv").
+- With this list (the 'positive' list) we used the fuzzy match algorithm to find a list (the 'negative' list) of the second most similar references to them (i.e. the best match would be the reference in question, and the second would be quite similar).
+- The positive list was reduced to 199 since one reference didn't have a second strong match, and the negative list gave 202 references since for some references there were multiple references with the same cosine similarity competing for second place.
+- We take the titles and publication IDs from these and combine them in "positive_and_negative_match_test_data.csv", and in this create a 'Do we expect a match?' column where we record a '1' if this reference was from the positive list and a '0' if it was from the negative list.
+
+### Evaluation Score 5
+- We remove the references with 'Do we expect a match?' = 0 from "uber_api_publications.csv", this is our 'publication set'.
```
Comment (nsorros) [inline docs/evaluation.md]: Ok now it is clearer

**Comment by nsorros** [comment]
Now I am getting `FileNotFoundError: [Errno 2] No such file or directory: './algo_evaluation/data_evaluate/pdfs/2713.pdf.xml.pdf'`

**Comment by nsorros** [comment]
I deleted all the files and redownloaded them and now I am getting

`KeyError: 'Do we expect a match?'`
